lib/ide/src/view/view_layout.rs

```diff
@@ -0,0 +1,159 @@
+    dimensions       : Vector2<f32>
+}
+
+impl Drop for ViewLayoutData {
+    fn drop(&mut self) {
+        if let Some(keyboard_closure) = self.keyboard_closure.as_ref() {
+            let body = document().unwrap().body().unwrap();
+            let callback : &Function = keyboard_closure.as_ref().unchecked_ref();
+            body.remove_event_listener_with_callback("keydown", callback).ok();
+        }
+    }
+}
+
+impl ViewLayoutData {
+    fn set_view_mode(&mut self, view_mode:ViewMode) {
+        self.view_mode = view_mode;
+        self.recalculate_layout();
+    }
+
+    fn switch_mode(&mut self) {
+        if let ViewMode::Half = self.view_mode {
+            self.set_view_mode(ViewMode::Full)
+        } else {
+            self.set_view_mode(ViewMode::Half)
+        }
+    }
+
+    fn set_dimensions(&mut self, dimensions:Vector2<f32>) {
+        self.dimensions = dimensions;
+        self.recalculate_layout();
+    }
+
+    fn recalculate_layout(&mut self) {
+        let dimensions = self.dimensions;
+        let (position,dimensions) = match self.view_mode {
+            ViewMode::Full => {
+                let position   = Vector2::new(0.0, dimensions.y);
+                (position,dimensions)
+            },
+            ViewMode::Half => {
+                let position   = Vector2::new(0.0, dimensions.y / 2.0);
+                let dimensions = Vector2::new(dimensions.x, dimensions.y / 2.0);
+                (position,dimensions)
+            }
+        };
+        let padding = Padding {
+            left   : 10.0,
+            top    : 10.0,
+            right  : 10.0,
+            bottom : 10.0
+        };
+        self.text_editor.set_padding(padding);
+        self.text_editor.set_dimensions(dimensions);
+        self.text_editor.set_position(position);
+        self.text_editor.update();
+    }
+}
+
+
+
+// ==================
+// === ViewLayout ===
+// ==================
+
+/// Initial implementation of ViewLayout with a single TextEditor. Pressing ctrl+f toggles
+/// fullscreen mode.
+#[derive(Debug,Clone)]
+pub struct ViewLayout {
+    data : Rc<RefCell<ViewLayoutData>>
+}
+
+impl ViewLayout {
+    /// Creates a new ViewLayout with a single TextEditor.
+    pub fn default(world:&World) -> Self {
+        let text_editor      = TextEditor::new(&world);
+        let keyboard_closure = None;
+        let view_mode        = ViewMode::Half;
+        let dimensions       = zero();
+        let data             = ViewLayoutData {text_editor,keyboard_closure,view_mode,dimensions};
+        let data             = Rc::new(RefCell::new(data));
+        Self {data}.init(world)
+    }
+
+    fn init_keyboard(self) -> Self {
+        let data    = Rc::downgrade(&self.data);
+        let closure = move |event:KeyboardEvent| {
+            const F_KEY : u32 = 70;
+            if event.ctrl_key() && event.key_code() == F_KEY {
+                if let Some(data) = data.upgrade() {
+                    data.borrow_mut().switch_mode()
```

Comment:
Lets try not to use `borrow_mut` in bigger code chunks. This should be `data.switch_mode()`. Btw use `shared!` to define `ViewLayout` and this method will be autogenerated :)